File: server/bbz_core/infra/models/telephony.py
"""Telephony core objects: lines, calls, participants, call documentation (E11-01).

MASTER_PROMPT §14, §8.4, §13.8/§13.10. Vendor-neutral — no Cisco/JTAPI types
here; a provider translates its events into the normalised shape
(``packages/event-schemas/telephony_event.v1.json``) and the call aggregate
(E11-04) drives the state.

Notes:

* every call has a BBZ-owned ``bbz_call_id`` that is **independent** of the
  provider's ``source_call_id`` (which may be null early, and is not globally
  unique);
* ``call_participants.number`` and ``call_documentation.free_text`` are
  personally identifiable — scope/retention applies (see
  ``docs/domain/retention-policy.md``);
* ``category`` is a closed CHECK set (§13.10), nullable until the operator sets
  it (the hangup guard, E11-10, keys off ``mandatory_done``).
"""

from __future__ import annotations

import datetime as _dt
import enum
import uuid

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    ForeignKey,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column

from bbz_core.infra.models.base import Base, TimestampMixin, uuid_pk


class LineState(enum.StrEnum):
    IN_SERVICE = "in_service"
    OUT_OF_SERVICE = "out_of_service"
    UNKNOWN = "unknown"


class CallDirection(enum.StrEnum):
    INBOUND = "inbound"
    OUTBOUND = "outbound"


class CallState(enum.StrEnum):
    OFFERED = "offered"
    RINGING = "ringing"
    CONNECTED = "connected"
    HELD = "held"
    TRANSFERRING = "transferring"
    DISCONNECTED = "disconnected"
    FAILED = "failed"
    #: hung up but not yet documented — stays "open" until a category is set
    #: (the mandatory-documentation guard, E11-10)
    ENDED_PENDING_DOCUMENTATION = "ended_pending_documentation"


class ParticipantRole(enum.StrEnum):
    CALLER = "caller"
    CALLEE = "callee"
    OPERATOR = "operator"
    TRANSFER_TARGET = "transfer_target"
    CONFERENCE = "conference"


class CallCategory(enum.StrEnum):
    #: §13.10: Auskunftsersuchen / Technische Störung / Reinigungsmeldung Kunde /
    #: EVU & EVI Mitteilung / Anderes
    INFORMATION_REQUEST = "information_request"
    TECHNICAL_FAULT = "technical_fault"
    CLEANING_REPORT_CUSTOMER = "cleaning_report_customer"
    EVU_EVI_NOTICE = "evu_evi_notice"
    OTHER = "other"


def _in(column: str, values: type[enum.StrEnum], name: str) -> CheckConstraint:
    allowed = ", ".join(f"'{v.value}'" for v in values)
    return CheckConstraint(f"{column} IN ({allowed})", name=name)


class Line(Base, TimestampMixin):
    __tablename__ = "lines"
    __table_args__ = (
        _in("state", LineState, "line_state"),
        UniqueConstraint("provider", "external_id", name="uq_lines_provider_external"),
    )

    id: Mapped[uuid.UUID] = uuid_pk()
    provider: Mapped[str] = mapped_column(String(64), index=True)
    external_id: Mapped[str] = mapped_column(String(128))
    label: Mapped[str | None] = mapped_column(String(200))
    state: Mapped[str] = mapped_column(String(20), server_default=LineState.UNKNOWN.value)
    workplace_id: Mapped[uuid.UUID | None] = mapped_column(index=True)


class Call(Base, TimestampMixin):
    __tablename__ = "calls"
    __table_args__ = (
        _in("direction", CallDirection, "call_direction"),
        _in("state", CallState, "call_state"),
        CheckConstraint(
            "caller_priority IS NULL OR caller_priority IN ('low', 'medium', 'high')",
            name="call_caller_priority",
        ),
    )

    id: Mapped[uuid.UUID] = uuid_pk()
    #: BBZ-owned, stable, human-scannable id — never the provider's
    bbz_call_id: Mapped[str] = mapped_column(String(32), unique=True)
    provider: Mapped[str] = mapped_column(String(64), index=True)
    #: the provider's call id — null until assigned, not globally unique
    source_call_id: Mapped[str | None] = mapped_column(String(128), index=True)
    direction: Mapped[str] = mapped_column(String(16))
    state: Mapped[str] = mapped_column(
        String(32), server_default=CallState.OFFERED.value, index=True
    )
    line_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("lines.id", ondelete="SET NULL"), index=True
    )
    workplace_id: Mapped[uuid.UUID | None] = mapped_column(index=True)
    started_at: Mapped[_dt.datetime | None] = mapped_column()
    ended_at: Mapped[_dt.datetime | None] = mapped_column()
    #: caller resolution (E11-08) — snapshotted at call time. NULL contact = the
    #: number did not resolve to a single contact ("unknown").
    caller_contact_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("contacts.id", ondelete="SET NULL"), index=True
    )
    caller_priority: Mapped[str | None] = mapped_column(String(16))


class CallParticipant(Base):
    __tablename__ = "call_participants"
    __table_args__ = (_in("role", ParticipantRole, "call_participant_role"),)

    id: Mapped[uuid.UUID] = uuid_pk()
    call_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("calls.id", ondelete="CASCADE"), index=True
    )
    number: Mapped[str | None] = mapped_column(String(64))
    display_name: Mapped[str | None] = mapped_column(String(200))
    role: Mapped[str] = mapped_column(String(20))
    created_at: Mapped[_dt.datetime] = mapped_column(server_default=text("now()"))


class CallDocumentation(Base):
    """One row per call (``call_id`` is the PK) — re-saving overwrites (E11-09)."""

    __tablename__ = "call_documentation"
    __table_args__ = (
        CheckConstraint(
            "category IS NULL OR category IN ("
            + ", ".join(f"'{c.value}'" for c in CallCategory)
            + ")",
            name="call_documentation_category",
        ),
    )

    call_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("calls.id", ondelete="CASCADE"), primary_key=True
    )
    category: Mapped[str | None] = mapped_column(String(32))
    free_text: Mapped[str | None] = mapped_column(Text)
    documented_by: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL")
    )
    documented_at: Mapped[_dt.datetime | None] = mapped_column()
    #: true once a category is set — the hangup guard (E11-10) blocks final
    #: ``CALL_ENDED`` until this is true
    mandatory_done: Mapped[bool] = mapped_column(Boolean, server_default=text("false"))
    updated_at: Mapped[_dt.datetime] = mapped_column(
        server_default=text("now()"), onupdate=text("now()")
    )

File: apps/web/src/lib/telephony.ts
import { api } from '@/lib/apiClient';

/** Telephony / call-control API (Epic 11 — E11-06/07/09/11/12, MASTER_PROMPT §13.8–13.11). */

export type CallState =
  | 'offered'
  | 'ringing'
  | 'connected'
  | 'held'
  | 'transferring'
  | 'disconnected'
  | 'failed'
  | 'ended_pending_documentation';

export type CallDirection = 'inbound' | 'outbound';
export type CallerPriority = 'low' | 'medium' | 'high';

export type CallCategory =
  | 'information_request'
  | 'technical_fault'
  | 'cleaning_report_customer'
  | 'evu_evi_notice'
  | 'other';

export const CALL_CATEGORIES: CallCategory[] = [
  'information_request',
  'technical_fault',
  'cleaning_report_customer',
  'evu_evi_notice',
  'other',
];

export interface CallParticipant {
  number: string | null;
  display_name: string | null;
  role: string;
}

export interface Call {
  id: string;
  bbz_call_id: string;
  provider: string;
  direction: CallDirection;
  state: CallState;
  line_id: string | null;
  workplace_id: string | null;
  started_at: string | null;
  ended_at: string | null;
  created_at: string;
  category: CallCategory | null;
  has_free_text: boolean;
  caller_contact_id: string | null;
  caller_priority: CallerPriority | null;
  participants: CallParticipant[];
}

export interface CallPage {
  items: Call[];
  next_cursor: string | null;
}

export interface ControlResult {
  call_id: string | null;
  action: string;
  accepted: boolean;
  detail: string | null;
}

export interface CallDocumentation {
  call_id: string;
  category: CallCategory | null;
  free_text: string | null;
  documented_by: string | null;
  documented_at: string | null;
  mandatory_done: boolean;
}

export interface Line {
  id: string;
  provider: string;
  external_id: string;
  label: string | null;
  state: string;
  workplace_id: string | null;
  updated_at: string;
}

export const PRIORITY_RANK: Record<string, number> = { high: 0, medium: 1, low: 2, unknown: 3 };

/** the "other party" on a call, for display (caller on inbound, callee on outbound). */
export function otherParty(call: Call): string {
  const wantRole = call.direction === 'inbound' ? 'caller' : 'callee';
  const p = call.participants.find((x) => x.role === wantRole) ?? call.participants[0];
  return p?.display_name ?? p?.number ?? '—';
}

export const telephonyApi = {
  ringing: (signal?: AbortSignal) => api.get<CallPage>('/calls/ringing', { signal }),

  history: (
    params: { limit?: number; number?: string; direction?: CallDirection } = {},
    signal?: AbortSignal,
  ) => {
    const qs = new URLSearchParams();
    if (params.limit) qs.set('limit', String(params.limit));
    if (params.number) qs.set('number', params.number);
    if (params.direction) qs.set('direction', params.direction);
    const tail = qs.toString();
    return api.get<CallPage>(`/calls${tail ? `?${tail}` : ''}`, { signal });
  },

  lines: (signal?: AbortSignal) => api.get<{ lines: Line[] }>('/lines', { signal }),

  dial: (line_id: string, destination: string) =>
    api.post<{ accepted: boolean; detail: string | null }>('/calls/dial', { line_id, destination }),

  answer: (id: string) => api.post<ControlResult>(`/calls/${id}/answer`),
  hangup: (id: string) => api.post<ControlResult>(`/calls/${id}/hangup`),
  hold: (id: string) => api.post<ControlResult>(`/calls/${id}/hold`),
  resume: (id: string) => api.post<ControlResult>(`/calls/${id}/resume`),
  transfer: (id: string, destination: string) =>
    api.post<ControlResult>(`/calls/${id}/transfer`, { destination }),

  getDoc: (id: string, signal?: AbortSignal) =>
    api.get<CallDocumentation>(`/calls/${id}/documentation`, { signal }),
  putDoc: (id: string, body: { category: CallCategory | null; free_text: string | null }) =>
    api.put<CallDocumentation>(`/calls/${id}/documentation`, body),

  pendingDocs: (signal?: AbortSignal) =>
    api.get<{ calls: { call_id: string; bbz_call_id: string; direction: string; ended_at: string | null }[] }>(
      '/calls/pending-documentation',
      { signal },
    ),
};
